src/library/PiaFormat.ts

```diff
@@ -0,0 +1,94 @@
+import dayjs from "dayjs";
+
+enum PIATypesEnum {
+  piaString = "string",
+  piaDate = "date",
+  piaMonthYear = "monthyear",
+  piaInt = "int",
+}
+
+const piaLineParsers = {
+  "01": (contents: string) => {
+    const endPositions = {
+      socialSec: { start: 3, end: 11, type: PIATypesEnum.piaString },
+      /* 0 is male, 1 is female, but we don't use it, just store it */
+      sex: { start: 12, end: 12, type: PIATypesEnum.piaInt },
+      birthDate: { start: 13, end: 20, type: PIATypesEnum.piaDate },
+      /* birthMonth: [13,14], 
+              // birthDay: [15,16],
+              // birthYear: [17,20] */
+    };
+    return Object.keys(endPositions).map((piece) => ({
+      key: piece,
+      value: constructPieceType(
+        contents.slice(
+          endPositions[piece].start - 3,
+          endPositions[piece].end - 2
+        ),
+        endPositions[piece].type
+      ),
+    }));
+
+    return {};
+  },
+};
+
+const constructPieceType = (value: string, ourType: PIATypesEnum) => {
+  switch (ourType) {
+    case PIATypesEnum.piaDate:
+      if (value.length !== 8) {
+        console.warn("piaFormat date is invalid length");
+      }
+      /* dayjs will use local timezone */
+      /* starts in mmddyyyy, dayjs takes yyyy-mm-dd */
+      return dayjs(
+          [value.slice(4, 8), value.slice(0, 2), value.slice(2, 4)]
+            .join("-")
+          ).toDate();
+      break;
+    case PIATypesEnum.piaInt:
+      return new Number(value);
+      break;
+    case PIATypesEnum.piaString:
+    default:
+      return value;
+      break;
+  }
+  return;
+};
+
+const piaLineParser = (lineId: string, contents: string) =>
+  [lineId, (lineId === "01" && piaLineParsers[lineId](contents)) || contents];
+
+export class PiaFormat {
+  constructor(piaInput: string, fileName: string) {
+    this.piaAll = piaInput;
+    const lines = piaInput
+      .split("\n")
+      .map((n) => [n.slice(0, 2), n.slice(2)])
```

Comment:
I'd avoid the `n.slice(2)` because it makes it harder for us to follow the spec exactly (we have to do the `start - 3` and `end - 2` stuff). After this, we can just for-loop over the entire thing and build up a `PiaData` object, as well as an "everything else" string that contains all the lines we couldn't parse.
Something like this (syntax might be wrong):
```typescript
  const lines = piaInput.split("\n").map((line) => [line.slice(0, 2), line])

  let piaData: Partial<PiaData> = {}
  let everythingElse = ""
  lines.forEach(line => {
    const [lineCode, strLine] = line
    const parser = lineParsers[lineCode]
    if (parser) Object.assign(piaData, parser(strLine))
    else everythingElse = `${everythingElse}\n${strLine}`
  })
```

By the end of that, we'll have a nice `piaData` object with all the fields that were parsed, and an `everythingElse` string with all the lines that we couldn't!